# Intro to built-in data types and numpy arrays

First of all, activate the `psychopy` conda envirornment (from the Terminal or Anaconda Prompt: `conda activate psychopy`, or from the Anaconda navigator) and let's check that `numpy` is installed.

In [1]:
import numpy as np

## 1. Built-in Python data types

Text and numeric types:

In [2]:
text = "ciao"
type(text)

str

In [3]:
number = 3
type(number)

int

In [4]:
another_number = 3.2
type(another_number)

float

Booleans are truth statements:

In [5]:
number > 3

False

In [6]:
boolean = number < 3
boolean

False

In [7]:
type(boolean)

bool

Sequences and mappings:

- List is a collection which is ordered and changeable. Allows duplicate members.
- Tuple is a collection which is ordered and unchangeable. Allows duplicate members.
- Dictionary is a collection which is unordered and changeable. No duplicate members.

In [8]:
a_list = [1, 2, 3]
type(a_list)

list

In [9]:
a_tuple = (2, "b")
type(a_tuple)

tuple

In [10]:
a_dictionary = {"age": 25, "height": 180, "measurements": [8, 6, 15]}
type(a_dictionary)

dict

## 2. Lists

In [11]:
list_of_int = [1, 2, 3, 4]
list_of_int

[1, 2, 3, 4]

In [12]:
list_of_str = ['a', 'b', 'c', 'defgh']
list_of_str

['a', 'b', 'c', 'defgh']

In [13]:
list_of_bool = [True, False, True, True]
list_of_bool

[True, False, True, True]

In [14]:
list_of_float = [0.1, 1, 2.1, 3.4]
list_of_float

[0.1, 1, 2.1, 3.4]

In [15]:
list_of_int[0] # first element of a list

1

In [16]:
type(list_of_bool)

list

In [17]:
type(list_of_float[1]), type(list_of_float[2]) # list can contain different data types

(int, float)

In [18]:
len(list_of_float) # how many elements are in this list?

4

In [19]:
list_of_float*3 # stack the same list

[0.1, 1, 2.1, 3.4, 0.1, 1, 2.1, 3.4, 0.1, 1, 2.1, 3.4]

In [20]:
list_of_float + list_of_bool + list_of_float # stack different lists

[0.1, 1, 2.1, 3.4, True, False, True, True, 0.1, 1, 2.1, 3.4]

In [21]:
list_of_float # the original does not change!

[0.1, 1, 2.1, 3.4]

In [22]:
list_of_float.append(list_of_bool) # add elements at the end

In [23]:
list_of_float # the original changed!

[0.1, 1, 2.1, 3.4, [True, False, True, True]]

In [24]:
len(list_of_float) # we appended a list as an element

5

In [25]:
for i in list_of_str: # now we append every element one at a time
    list_of_int.append(i)

In [26]:
list_of_int

[1, 2, 3, 4, 'a', 'b', 'c', 'defgh']

In [27]:
list_of_int.sort() # you cannot sort the list when the elements are of different type though

TypeError: '<' not supported between instances of 'str' and 'int'

In [28]:
list_of_int = [2, 15, 6, 1]

In [29]:
list_of_int.sort()

In [30]:
list_of_int

[1, 2, 6, 15]

In [31]:
list_of_int.reverse() # reverse the order!
list_of_int

[15, 6, 2, 1]

In [32]:
list_of_int.remove(15) # remove a specific element
list_of_int

[6, 2, 1]

In [33]:
list_of_int.insert(2, 34) # insert a specific element in a certain position
list_of_int

[6, 2, 34, 1]

In [34]:
list_of_int.pop(2) # remove a specific element in a certain position
list_of_int

[6, 2, 1]

## 3. Dictionaries

In [35]:
auto = {'model': 'PEUGEOT 308',
        'CHF': 10500,
        'where': 'Dietikon',
        'km': 86000,
        'fuel': 'diesel',
        'ccm': 1560
       }

In [36]:
auto.values()

dict_values(['PEUGEOT 308', 10500, 'Dietikon', 86000, 'diesel', 1560])

In [37]:
auto.keys()

dict_keys(['model', 'CHF', 'where', 'km', 'fuel', 'ccm'])

In [38]:
auto.items()

dict_items([('model', 'PEUGEOT 308'), ('CHF', 10500), ('where', 'Dietikon'), ('km', 86000), ('fuel', 'diesel'), ('ccm', 1560)])

In [39]:
for i in auto.items():
    print(type(i))

<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>


In [40]:
for i in auto:
    print(i)

model
CHF
where
km
fuel
ccm


In [41]:
for i in auto:
    print(auto[i])

PEUGEOT 308
10500
Dietikon
86000
diesel
1560


In [42]:
for i in auto:
    print(i, auto[i])

model PEUGEOT 308
CHF 10500
where Dietikon
km 86000
fuel diesel
ccm 1560


In [43]:
len(auto) # how many items?

6

In [44]:
auto['model'] # get value corresponding to a key

'PEUGEOT 308'

In [45]:
auto['km'] = 90000 # change a value
auto

{'model': 'PEUGEOT 308',
 'CHF': 10500,
 'where': 'Dietikon',
 'km': 90000,
 'fuel': 'diesel',
 'ccm': 1560}

In [46]:
auto['color'] = 'blue' # add an item
auto

{'model': 'PEUGEOT 308',
 'CHF': 10500,
 'where': 'Dietikon',
 'km': 90000,
 'fuel': 'diesel',
 'ccm': 1560,
 'color': 'blue'}

In [47]:
'color' in auto # is a key in the dic?

True

In [48]:
auto.pop("CHF") # remove an item
auto

{'model': 'PEUGEOT 308',
 'where': 'Dietikon',
 'km': 90000,
 'fuel': 'diesel',
 'ccm': 1560,
 'color': 'blue'}

## 4. Creating a numpy array 

In [49]:
np.array() 

TypeError: Required argument 'object' (pos 1) not found

In [50]:
first_array = np.array([])
first_array

array([], dtype=float64)

In [51]:
int_array = np.array([1, 2, 3, 4])
int_array

array([1, 2, 3, 4])

In [52]:
int_array.dtype

dtype('int64')

In [92]:
str_array = np.array(['a', 'b', 'c', 'd', 'e'])
str_array.dtype

dtype('<U1')

In [96]:
str_array = np.array(['aerf', 'B1', 'c', 'd', 'e'])
str_array.dtype

dtype('<U4')

In [53]:
type(int_array), type(first_array), type(str_array)

(numpy.ndarray, numpy.ndarray)

In [54]:
int_array[0] # access the first element

1

In [74]:
more_dimensions = np.array([[1, 2, 3], [4.1, 5, 6]])
more_dimensions

array([[1. , 2. , 3. ],
       [4.1, 5. , 6. ]])

In [89]:
more_dimensions[0, 2] # access elements in a 2D array

3.0

In [90]:
more_dimensions[1, 0] = 4 # change an element
more_dimensions

array([[1., 2., 3.],
       [4., 5., 6.]])

In [75]:
more_dimensions.dtype # differently from lists, the elements of an array must all have same type

dtype('float64')

In [91]:
more_dimensions.astype(int) # change type

array([[1, 2, 3],
       [4, 5, 6]])

In [73]:
int_array.shape

(4,)

In [72]:
more_dimensions.shape

(2, 3)

Create arrays made of zeros or ones:

In [78]:
a = np.zeros(shape=(3, 10))
a

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [82]:
b = np.ones(10)
b

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [85]:
a*b # differently from lists, operations between arrays are math operations

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [86]:
a + b

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

In [87]:
b*24

array([24., 24., 24., 24., 24., 24., 24., 24., 24., 24.])

In [88]:
a + 10 / 2

array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5., 5., 5., 5., 5., 5.]])

Create different kind of sequences:

In [64]:
np.arange(start=0, stop=11, step=2) # sequence of numbers from start to stop, every step

array([ 0,  2,  4,  6,  8, 10])

In [66]:
np.arange(1, 12, step=2)

array([ 1,  3,  5,  7,  9, 11])

In [67]:
np.linspace(start=1, stop=10, num=10) # sequence of num lenghts, between start and stop

array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])

In [70]:
np.linspace(start=0, stop=10, num=21)

array([ 0. ,  0.5,  1. ,  1.5,  2. ,  2.5,  3. ,  3.5,  4. ,  4.5,  5. ,
        5.5,  6. ,  6.5,  7. ,  7.5,  8. ,  8.5,  9. ,  9.5, 10. ])

## 5. Boolean indexing

In [105]:
c = np.arange(1, 101, 1).reshape(10, 10)
c

array([[  1,   2,   3,   4,   5,   6,   7,   8,   9,  10],
       [ 11,  12,  13,  14,  15,  16,  17,  18,  19,  20],
       [ 21,  22,  23,  24,  25,  26,  27,  28,  29,  30],
       [ 31,  32,  33,  34,  35,  36,  37,  38,  39,  40],
       [ 41,  42,  43,  44,  45,  46,  47,  48,  49,  50],
       [ 51,  52,  53,  54,  55,  56,  57,  58,  59,  60],
       [ 61,  62,  63,  64,  65,  66,  67,  68,  69,  70],
       [ 71,  72,  73,  74,  75,  76,  77,  78,  79,  80],
       [ 81,  82,  83,  84,  85,  86,  87,  88,  89,  90],
       [ 91,  92,  93,  94,  95,  96,  97,  98,  99, 100]])

In [106]:
c > 45

array([[False, False, False, False, False, False, False, False, False,
        False],
       [False, False, False, False, False, False, False, False, False,
        False],
       [False, False, False, False, False, False, False, False, False,
        False],
       [False, False, False, False, False, False, False, False, False,
        False],
       [False, False, False, False, False,  True,  True,  True,  True,
         True],
       [ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True],
       [ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True],
       [ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True],
       [ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True],
       [ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True]])

In [108]:
c[c <= 10] = 0
c

array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
       [ 11,  12,  13,  14,  15,  16,  17,  18,  19,  20],
       [ 21,  22,  23,  24,  25,  26,  27,  28,  29,  30],
       [ 31,  32,  33,  34,  35,  36,  37,  38,  39,  40],
       [ 41,  42,  43,  44,  45,  46,  47,  48,  49,  50],
       [ 51,  52,  53,  54,  55,  56,  57,  58,  59,  60],
       [ 61,  62,  63,  64,  65,  66,  67,  68,  69,  70],
       [ 71,  72,  73,  74,  75,  76,  77,  78,  79,  80],
       [ 81,  82,  83,  84,  85,  86,  87,  88,  89,  90],
       [ 91,  92,  93,  94,  95,  96,  97,  98,  99, 100]])

## Exercises

- 1.1 Create a dictionary called `dog`, made of 5 meaningful items of your choice. Values should include: at least 1 integer, 1 list, 1 string.
- 1.2 Add an element to the list that you stored in the `dog` dictionary, and print the new dictionary.
- 1.3 Eliminate the item whose value was an integer, and print the new dictionary.
- 1.4 Add a new item, print the dictionary, and then delete it, and print it again.


- 2.1 Create a numpy array called `a`, of shape (10, 2), made of ones.
- 2.2 Create a numpy array called `b`, of shape (10, 1), made of zeros.
- 2.3 Add `a` to `b` and call their result `ab`.
- 2.4 Access the second element in the third row of `ab` and assign the value 100. Print the new `ab`.
- 2.5 Create a sequence called `c`, containing all multiples of 10 up to 200.
- 2.6 Create a sequence called `d`, containing all numbers from 0 to 2 with 2 decimals.

**Send me your solutions** before the end of this Friday (at laura.fontanesi@unibas.ch).

**NOTE**: Do not send me a notebook but simply a text file, named `wpa_2_Surname_Name.py` (with your own name and surname of course). To make it a bit clear, please write a comment to indicate which answer a specific chunk of code releated to, e.g.:

```
# exercise 1
a = 7 + 8
b = 8
c = a*b

# exercise 2
e = np.array([4, 5, 76])
f = np.append(e, [3, 4, 5])
```